### PR TITLE
fillchars: adding foldopen, foldsep, foldclose

### PR DESCRIPTION
**Pull request by teto**
You can try it with
set fillchars+=foldopen:▾,foldsep:│,foldclose:▸

should work with right/left too.

### CODE REVIEW

File: src/nvim/screen.c [hunk 1]
```diff
@@ -2017,58 +2016,73 @@ static void copy_text_attr(int off, char_u *buf, int len, int attr)
...
+    if (symbol != 0) {
+      // rollback length
+      char_counter -= len;
+    }
+    symbol = wp->w_p_fcs_chars.foldclosed;
+    len = utf_char2bytes(symbol, &p[char_counter]);
+    char_counter += len;
   }
+
+  return MAX(char_counter + (fdc-i), fdc);
```
Comment (norcalli) [inline src/nvim/screen.c]: This is the line that caused the build to fail because it is a comparison between an `unsigned long` and `int`

Comment (teto) [reply src/nvim/screen.c]: thanks. Fixed

**Comment by blueyed** [comment]
@teto 
To me this feels like it should be suggested to Vim already / also, to not cause unnecessary differences if it gets done there eventually also in a different way.
Is there an upstream issue for this feature, or some other discussion already?

**Comment by chrisbra** [comment]
see: https://github.com/vim/vim/issues/5206

**Comment by justinmk** [comment]
> To me this feels like it should be suggested to Vim already / also, to not cause unnecessary differences if it gets done there eventually also in a different way.

This hasn't been a problem for other fillchars that we've added, and it shouldn't block us for doing so here.

File: test/functional/ui/fold_spec.lua [hunk 2]
```diff
@@ -60,6 +60,47 @@ describe("folded lines", function()
     ]])
   end)
 
+  it("works with multibyte fillchars", function()
+    -- Currently the only allowed value of 'maxcombine'
```
Comment (justinmk) [inline test/functional/ui/fold_spec.lua]: ?

File: test/functional/ui/fold_spec.lua [hunk 3]
```diff
@@ -60,6 +60,47 @@ describe("folded lines", function()
...
+      {7:││}cc                                         |
+      {7:││}dd                                         |
+      {7:││}ee                                         |
+      {7:│ }ff                                         |
+      {1:~                                            }|
+      :1                                           |
+    ]]}
+
+    feed_command("set rightleft")
+screen:expect{grid=[[
```
Comment (justinmk) [inline test/functional/ui/fold_spec.lua]: indent

File: src/nvim/screen.c [hunk 4]
```diff
@@ -2017,58 +2016,73 @@ static void copy_text_attr(int off, char_u *buf, int len, int attr)
...
- * Only to be called when 'foldcolumn' > 0.
- */
-static void
-fill_foldcolumn (
+/// Fills the foldcolumn at "p" for window "wp".
+/// Only to be called when 'foldcolumn' > 0.
+///
+/// @param[out] p  Char array to write into
+/// @param lnum    Absolute current line number
+/// @param closed  Wether it is in foldcolumn mode
```
Comment (justinmk) [inline src/nvim/screen.c]:
```suggestion
/// @param closed  Whether it is in 'foldcolumn' mode
```

File: src/nvim/screen.c [hunk 5]
```diff
@@ -2017,58 +2016,73 @@ static void copy_text_attr(int off, char_u *buf, int len, int attr)
...
-  int empty;
-  int fdc = compute_foldcolumn(wp, 0);
-
+  int fdc = compute_foldcolumn(wp, 0);    // available cell width
+  size_t char_counter = 0;
+  int symbol = 0;
+  int len = 0;
   // Init to all spaces.
-  memset(p, ' ', (size_t)fdc);
+  memset(p, ' ', 6 * fdc + 1);
```
Comment (justinmk) [inline src/nvim/screen.c]: use `MAX_MCO` instead of magic 6 (same for other uses of 6 below)?

File: src/nvim/buffer_defs.h [hunk 6]
```diff
@@ -1120,6 +1120,9 @@ struct window_S {
     int stlnc;
     int vert;
     int fold;
+    int foldopen;                    /// when fold is open
+    int foldclosed;                  /// when fold is closed
+    int foldsep;                     /// continuous fold marker
```
Comment (justinmk) [inline src/nvim/buffer_defs.h]: use `///<` as in other cases in this same file

**Review by justinmk — APPROVED** [review]
lgtm after comments

File: test/functional/ui/fold_spec.lua [hunk 7]
```diff
@@ -60,6 +61,45 @@ describe("folded lines", function()
...
+      {7:││}cc                                         |
+      {7:││}dd                                         |
+      {7:││}ee                                         |
+      {7:│ }ff                                         |
+      {1:~                                            }|
+      :1                                           |
+    ]]}
+
+    feed_command("set rightleft")
+    screen:expect{grid=dedent([[
```
Comment (justinmk) [inline test/functional/ui/fold_spec.lua]: screen:expect() already dedents.

**Pull request merged**